CDAP-1763 fileset upgrade

rohitsinha54 opened the pull request:
Upgraded Filesets base paths:
Issue: https://issues.cask.co/browse/CDAP-1763

This PR was reviewed here: #1576

Opening this against develop and waiting for final approval.

Build : https://builds.cask.co/browse/CDAP-DUT1067-1

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetInstanceMDSUpgrader.java:
```diff
@@ -102,6 +106,43 @@ public Void apply(UpgradeMDSStores<DatasetInstanceMDS> ctx) throws Exception {
+        List<DatasetSpecification> dsSpecs = ctx.getNewMds().list(key, DatasetSpecification.class);
+        List<DatasetSpecification> fileSetSpecs = Lists.newArrayList();
+        for (DatasetSpecification dsSpec : dsSpecs) {
+          if (isFileSet(dsSpec)) {
```
this finds the FileSets that are registered as instances. What about FileSets that are embedded in other datasets? You would have to recurse through the embedded specs of each dataset to find them.

rohitsinha54 replied:
Yes. In the same way we upgrade their instance meta.

anew replied:
Consider where I have a dataset type WithFiles that itself is not a FileSet but has an embedded FileSet. For that dataset, isFileSet will return false, and so you will exclude it from the list of specs returned by this method. That means the embedded FileSet will never be inspected. You need to fix this.

rohitsinha54 replied:
Oh I didn't know that was possible too. Okay doing it now.

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
+   * @throws IOException
+   */
+  private void upgradeFileSet(DatasetSpecification dsSpec) throws IOException {
+    for (DatasetSpecification embeddedDsSpec : dsSpec.getSpecifications().values()) {
```
this only gets called for FileSets. They never have embedded datasets.

rohitsinha54 replied:
Please see below.

anew replied:
Here you do not need to iterate over the embedded specs.

rohitsinha54 replied:
Yes it was before I knew that a dataset which is not a fileset can have embedded filesets

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
+   */
+  private void upgradeFileSet(DatasetSpecification dsSpec) throws IOException {
+    for (DatasetSpecification embeddedDsSpec : dsSpec.getSpecifications().values()) {
+      upgradeFileSet(embeddedDsSpec);
```
And you cannot assume that every embedded dataset is a FileSet. Does the recursive call fail if you call this on a dataset that is not a FileSet?

rohitsinha54 replied:
I was not aware of this case. Thanks. Will fix it.

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
+    for (DatasetSpecification embeddedDsSpec : dsSpec.getSpecifications().values()) {
+      upgradeFileSet(embeddedDsSpec);
+    }
+    String basePath = FileSetProperties.getBasePath(dsSpec.getProperties());
```
Better check that the basePath is not null.

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
+    Location newlocation = locationFactory.create(Constants.DEFAULT_NAMESPACE)
+      .append(cConf.get(Constants.Dataset.DATA_DIR)).append(basePath);
+    LOG.info("Upgrading base path for dataset {} from {} to {}", dsSpec.getName(), oldLocation, newlocation);
+    renameLocation(oldLocation, newlocation);
```
I am not sure where this method went. In the last PR it was there and there was quite some discussion around it. We need to make sure that:
- calling this for a dataset that has already been upgraded does not fail
- calling this for multiple datasets that have the same base path does not cause failure

rohitsinha54 replied:
Since I moved this class that method was not needed anymore. Its in upgrade tool and it can use the common shared method which I have already changed according to your suggestions. 
Here is the PR which got merged: https://github.com/caskdata/cdap/pull/1580

rohitsinha54 replied:
No we use rename here: 
https://github.com/caskdata/cdap/blob/fa44c400e335b13e00a495f1f7ee91af3c3e091c/cdap-master/src/main/java/co/cask/cdap/data/tools/AbstractUpgrader.java#L57

So, if the basePath got moved initially then for other datasets it will say that it might have been moved and will not throw does FileNotFoundException as the renameTo does as we skip rename if it doesn't. So, I think that should be fine.


But I haven't tested the FileSet upgrade for FileSets which have same basepath. That is the only thing which I have on my plate for today. I have tested it with FileSet example.

anew commented:
@rohitsinha54 this still need some fixing. Please see my comments.

rohitsinha54 commented:
@anew : Addressed comment. Please have a look. Thanks

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetInstanceMDSUpgrader.java:
```diff
@@ -102,6 +106,43 @@ public Void apply(UpgradeMDSStores<DatasetInstanceMDS> ctx) throws Exception {
+        MDSKey key = new MDSKey(Bytes.toBytes(DatasetInstanceMDS.INSTANCE_PREFIX));
+        List<DatasetSpecification> dsSpecs = ctx.getNewMds().list(key, DatasetSpecification.class);
+        List<DatasetSpecification> fileSetSpecs = Lists.newArrayList();
+        for (DatasetSpecification dsSpec : dsSpecs) {
```
In theory, an embedded dataset can embed another - it is recursive. I don't think it is common, but you shuold do this recursively, too.

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +100,33 @@ public void upgrade() throws Exception {
+        upgradeFileSet(embeddedDsSpec);
+      }
+    }
+    String basePath = FileSetProperties.getBasePath(dsSpec.getProperties());
```
Make sure that basePath is not null

anew commented:
Please address my comments. I feel like I am repeating the same comments.

rohitsinha54 commented:
@anew : Sorry about that. I thought only a dataset which is fileset can have embedded fileset. Didn't know any dataset can have embedded filesets in it. Doing it now.

rohitsinha54 commented:
@anew : Addressed all the comments and also went through all previous comments to make sure I addressed everything as a double check :)
My initial understanding for where all can I find filesets was wrong which cause a bit of problem. Thanks for making it clear to me. I have fixed it. Please have a look.
Thanks.

anew commented on cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java:
```diff
@@ -96,6 +101,28 @@ public void upgrade() throws Exception {
+   */
+  private void upgradeFileSet(DatasetSpecification dsSpec) throws IOException {
+    String basePath = FileSetProperties.getBasePath(dsSpec.getProperties());
+    Preconditions.checkNotNull(basePath, "The basepath for fileset {} is null", dsSpec.getName());
```
What I meant was that the upgrade tool would not fail. The Precondition only makes it throw the NPE one line earlier. Can you just ignore the fileset if it does not have a base path?

rohitsinha54 replied:
Will do. But quick question. Is it possible to have a fileset which does not have a basepath ?

anew commented:
Only one comment and you are good to merge

rohitsinha54 commented:
Build :green_heart: http://builds.cask.co/browse/CDAP-DUT1067-6
Merging.
